aiida_common_workflows/common/init_magnetization.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+"""
+Utilities to manage initial magnetization.
+"""
+from aiida.orm import StructureData
+
+
+class InitMagnetizationManager():
+    """
+    A manager to facilitate the management of initial magnetizations.
+    The API agreed in the project impose to pass the initial magnetizations
+    as a list with the spin polarization (difference between spin up and down electrons)
+    for each site.
+    The manager requires in input for the initaialization this list and the structure
+    and contains methods to manage these quantities.
+    """
+
+    def __init__(self, structure, initial_spin_site):
+        """
+        Construct an instance of InitMagnetizationManager
+        :param: structure: a StructureData
+        :param: initial_spin_site: a list with the spin polarization per site.
+                The values are the difference between the up-spin electrons and
+                down-spin electrons in the site. The order of the list follows the
+                one of the structure.sites
+        """
+
+        if not isinstance(structure, StructureData):
+            raise RuntimeError('structure input must be a StructureData')
+
+        if not isinstance(initial_spin_site, list):
+            raise RuntimeError('initial_spin_site must be a list')
+
+        if len(initial_spin_site) != len(structure.sites):
+            raise RuntimeError('an initial magnetization for each site must be spedified')
+
+        self.structure = structure
+        self.magn_list = initial_spin_site
+
+    def get_per_site_list(self):
+        """
+        Return the list of spin polarization per site.
+        """
+        return self.magn_list
+
+    def get_per_kind_dict_and_structure(self):
```

Comment:
Two high-level comments for this function before I go into the details of the implementation:

* You are potentially returning a new structure, but it is not going through a `calcfunction` so you will lose the provenance.
* If I understand correctly, you are doing exact comparisons on the magnetization for each site to see if they are the same and so can be modeled by the same kind. Would it be sensible to include some tolerances here? i.e. if you have 0.5 and 0.5000001, I think they can safely be considered to be the same kind

On the last point, I have actually been working on these very problems and already implemented a `calcfunction` that does this for Quantum ESPRESSO, but that should be generalizable.

On that note, is this manager actually being used anywhere in the code now? Or is the idea that the plugins can use them should they want to?

Reply:
>     * You are potentially returning a new structure, but it is not going through a `calcfunction` so you will lose the provenance.

The idea is that is called inside `get_builder`, therefore is not such a big deal to lose the provenance, actually everything that happens in `get_builder` is lost. Only the inputs of the workchain remain.


>     * If I understand correctly, you are doing exact comparisons on the magnetization for each site to see if they are the same and so can be modeled by the same kind. Would it be sensible to include some tolerances here? i.e. if you have 0.5 and 0.5000001, I think they can safely be considered to be the same kind

Good suggestion


> On the last point, I have actually been working on these very problems and already implemented a `calcfunction` that does this for Quantum ESPRESSO, but that should be generalizable.


Fell free to replace with your implementation

> On that note, is this manager actually being used anywhere in the code now? Or is the idea that the plugins can use them should they want to?

The idea is to make it available for developers, if they want to use it.

Reply:
I think it is intended for the code like Quantum ESPRESSO that accepts the `{'kind_name': spin}` type input. So the plugins decide if they want it so not.

Another comment - yes the creation of a new `StructureData` should be wrapped in a `calcfunction`. But I think that the creation of a new structure may not be necessary if the input structure already has the required types matching the list of magnetic moments. Perhaps this can be moved to a standalone function and the unwrapped version can be called first - if the new `StructureData` contains new kinds then we actually call the wrapped version - otherwise just return the input `StructureData`. 

Yes if QuantumESPRESSO plugin has it implemeted we can steal and generalizble it from there.

Reply:
> The idea is that is called inside `get_builder`, therefore is not such a big deal to lose the provenance, actually everything that happens in `get_builder` is lost.

Yes, but the `structure` is a real exception. While all other inputs of the builder are normal non-node inputs, the structure is an actual node. If you pass it as the structure to relax, and the builder silently creates a new structure without a link, then the relaxed structure and the rest of the results of the workchain are not linked to the structure you passed in. This is definitely not what we want and not what a user would expect.

Reply:
@bosonie  Since this code is not actually used or tested in this PR yet, and it might need some more discussion, would it make sense to take it out of this PR so we can get the rest merged and available for the plugin developers?

Reply:
yes